src/ngx_http_lua_log_ringbuf.c

```diff
@@ -0,0 +1,206 @@
+#endif
+#include "ddebug.h"
+
+
+#include "ngx_http_lua_common.h"
+#include "ngx_http_lua_log_ringbuf.h"
+
+
+typedef struct {
+    unsigned    is_data:1;
+    unsigned    log_level:4;
+    int         len;
+} ngx_http_lua_log_ringbuf_header_t;
+
+
+static void * ngx_http_lua_log_ringbuf_next_header(
+    ngx_http_lua_log_ringbuf_t *rb);
+static void ngx_http_lua_log_ringbuf_append(
+    ngx_http_lua_log_ringbuf_t *rb, int is_data, int log_level, void *buf,
+    int n);
+static size_t ngx_http_lua_log_ringbuf_free_spaces(
+    ngx_http_lua_log_ringbuf_t *rb);
+
+
+void
+ngx_http_lua_log_ringbuf_init(ngx_http_lua_log_ringbuf_t *rb, void *buf,
+    size_t len)
+{
+    rb->data = buf;
+    rb->size = len;
+
+    rb->tail = rb->data;
+    rb->head = rb->data;
+    rb->count = 0;
+    rb->filter_level = NGX_LOG_DEBUG;
+
+    return;
+}
+
+
+void
+ngx_http_lua_log_ringbuf_reset(ngx_http_lua_log_ringbuf_t *rb)
+{
+    rb->tail = rb->data;
+    rb->head = rb->data;
+    rb->count = 0;
+
+    return;
+}
+
+
+/*
+ * get the next data header, it'll skip the useless data space or
+ * placehold data
+ */
+
+static void *
+ngx_http_lua_log_ringbuf_next_header(ngx_http_lua_log_ringbuf_t *rb)
+{
+    ngx_http_lua_log_ringbuf_header_t       *head;
+
+    /* useless data */
+    if (rb->size - (rb->head - rb->data) <
+        sizeof(ngx_http_lua_log_ringbuf_header_t))
+    {
+        rb->head = rb->data;
+        return rb->head;
+    }
+
+    /* placehold data */
+    head = (ngx_http_lua_log_ringbuf_header_t *) rb->head;
+    if (!head->is_data) {
+        rb->head = rb->data;
+        return rb->head;
+    }
+
+    return rb->head;
+}
+
+
+/* append data to ring buffer directly */
+
+static void
+ngx_http_lua_log_ringbuf_append(ngx_http_lua_log_ringbuf_t *rb,
+    int is_data, int log_level, void *buf, int n)
+{
+    ngx_http_lua_log_ringbuf_header_t        *head;
+
+    head = (ngx_http_lua_log_ringbuf_header_t *) rb->tail;
+    head->len = n;
+    head->log_level = log_level;
+    head->is_data = is_data;
+
+    if (!is_data) {
+        rb->tail = rb->data;
+        return;
+    }
+
+    rb->tail += sizeof(ngx_http_lua_log_ringbuf_header_t);
+    ngx_memcpy(rb->tail, buf, n);
+    rb->tail += n;
+    rb->count++;
+
+    return;
+}
+
+
+/* size of free spaces */
+
+static size_t
+ngx_http_lua_log_ringbuf_free_spaces(ngx_http_lua_log_ringbuf_t *rb)
+{
+    if (rb->tail == rb->head && rb->tail == rb->data) {
+        return rb->size;
+    }
+
+    if (rb->tail > rb->head) {
+        return rb->data + rb->size - rb->tail;
+    }
+
+    return rb->head - rb->tail;
+}
+
+
+/*
+ * try to write log data to ring buffer, throw away old data
+ * if there was not enough free spaces.
+ */
+
+ngx_int_t
+ngx_http_lua_log_ringbuf_write(ngx_http_lua_log_ringbuf_t *rb, int log_level,
+    void *buf, size_t n)
+{
+    size_t              free_spaces, head_len;
+
+    ngx_http_lua_log_ringbuf_header_t       *head;
+
+    head_len = sizeof(ngx_http_lua_log_ringbuf_header_t);
```

Comment:
We shouldn't use an automatic variable for constant values like this. It is best practice to use anonymous enum for this:

```C
enum {
    HEADER_LEN = sizeof(ngx_http_lua_log_ringbuf_header_t),
};
```

Also, we should use "header len" instead of "head len". The latter is confusing.